_posts/2018-02-12-Reusable-features-in-React-Redux.md

```diff
@@ -0,0 +1,188 @@
+
+
+#### selectors.js
+
+The file where you define your selectors. Selectors are pure functions that compose and know how
+to retrieve data from your state. If you're now familiar with the concept I highly recommend looking 
+at [reselect](https://github.com/reactjs/reselect) library. I strongly recommend dividing the selectors into two categories. 
+Input-selectors that only retrieve data from the state and combined-selectors that can combine
+multiple input-selectors and add some business logic or do some transformations.
+
+
+#### utils.js
+
+Everything related to your feature that doesn't belong anywhere else. Usually it contains support utilities and helpers.
+
+#### middleware.js
+
+This is where our feature handles it's side effects, e.g. catching some Redux actions and dispatching new ones,
+synchronize the API calls, etc...
+
+You can check the [reference repository](https://github.com/char0n/react-reusable-features/tree/master/src/app/standard-feature) to see how this looks in a real project. 
+It depends on your use-case what the standard feature directory contains. 
+There are situations that you may need reducers but don't need actions and so on...
+But regardless, the full representation of the standard feature should look like that. 
+If the number of the components in the standard feature becomes too big, you break the standard
+feature into more sub-features which also adheres to the same rules as it's parent feature.     
+
+
+```
+src/
+ todo-feature/
+   components/ 
+     TodoList.js
+   containers/
+     TodoList.js
+   todo-item/
+     components/
+       TodoListItem.js
+     containers/
+       TodoListItem.js
+     actions.js
+     selectors.js 
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```
+
+
+It is always good to break features into smaller sub-features with encapsulated logic to reduce the complexity.
+Try to think of it as applying Single Responsibility Principle. One feature does only one thing and
+it's entire logic is isolated in it's directory. As you applications grows and you add more and more features, 
+it really pays of to structure your code using these rules. Somebody that will come to work on a project after
+you will really thank you ;]
+
+The standard feature has unfortunately one disadvantage. It is statically mounted on specific place in a state. 
+Imagine a situation when you want to display two Todo Lists on the same page.
+If you do that, it will work and the Todo Lists will display. The problem arises when you try to interact with them. 
+Changes done in one list (checking the item as done, reorder...) will be reproduced also on the second one.
+The cause of this effect is the fact that they are both focused on the same place in the Redux state and 
+act accordingly on Redux state changes. You may never intercept this situation, it really depends
```

Comment:
Intercept isn't the right word here, I cannot recall which word you mean though.

Perhaps as an alternative you could put "experience".

```diff
- You may never intercept this situation,
+ You may never experience this situation,
```